Phpstan 0.10 compat

Jean85 opened the pull request:
This PR includes #4, and adds the right composer.json constraint.

localheinz commented:
@Jean85 

Note that `composer.lock` is checked in.

localheinz commented:
@Jean85 

Also note that the build for PHP 7.0 needs to be dropped from `.travis.yml`, and also see #13.

Jean85 commented:
Thanks @localheinz, I've addressed the Travis issue and I didn't touch the composer.lock, we can address that with your PR.

I've also refactored the classes, because there were others BCs. The solution is inspired by the code of the phpstan-phpunit extension: https://github.com/phpstan/phpstan-phpunit/blob/master/src/Type/PHPUnit/GetMockBuilderDynamicReturnTypeExtension.php

Jan0707 commented:
Dropping Support for PHP7.0 would require us to move to version 0.2
Should me make this a separate branch in order to maintain a version for PHP7.0 ?

Jean85 commented:
@Jan0707 yes, but I would advise you to do the opposite: you should branch from the last tagged commit in the 0.1.x series.

Jan0707 commented:
So, I guess the plan is as follows: Once #9 is resolved, I will release a version (last one with PHP7.0 support) and then we can finish this issue and release the next version.

Jan0707 commented:
I have just released [0.1.2](https://github.com/Jan0707/phpstan-prophecy/releases/tag/0.1.2) :)

So I guess we can now work on this issue.

Jean85 commented on .travis.yml:
```diff
@@ -36,30 +36,6 @@ jobs:
-
-      php: 7.1
-
-      env: WITH_LOCKED=true
```
I've removed this one because the `composer.lock` file is generated under PHP 7.2, so it should be used only under the same platform.

localheinz replied:
@Jean85 

Ha, I think it would be best to check in the version of `composer.lock` as generated on the lowest supported version, what do you think?

Jean85 replied:
No opinions about that for me! I would leave it to @Jan0707

Jan0707 replied:
Still not sure whether to actually check it in at all. I did initially check it in to make sure all my dev machines were running the same versions of dependencies, but I also see the downsides.

Jean85 replied:
I would leave it like that and postpone the decision then, it's better if we release this PR ASAP and measure any downside in practice, if any.
